packages/core/lib/asciidoctor/js/asciidoctor_ext/node/template.rb

```diff
@@ -0,0 +1,194 @@
+      # FIXME need to think about safe mode restrictions here
+      next unless ::File.directory?(template_dir = (path_resolver.system_path template_dir))
+
+      if engine
+        file_pattern = %(*.#{engine})
+        # example: templates/haml
+        if ::File.directory?(engine_dir = %(#{template_dir}/#{engine}))
+          template_dir = engine_dir
+        end
+      else
+        # NOTE last matching template wins for template name if no engine is given
+        file_pattern = '*'
+      end
+
+      # example: templates/html5 (engine not set) or templates/haml/html5 (engine set)
+      if ::File.directory?(backend_dir = %(#{template_dir}/#{backend}))
+        template_dir = backend_dir
+      end
+
+      pattern = %(#{template_dir}/#{file_pattern})
+
+      if (scan_cache = @caches[:scans])
+        template_cache = @caches[:templates]
+        unless (templates = scan_cache[pattern])
+          templates = scan_cache[pattern] = scan_dir template_dir, pattern, template_cache
+        end
+        templates.each do |name, template|
+          @templates[name] = template_cache[template.file] = template
+        end
+      else
+        @templates.update scan_dir(template_dir, pattern, @caches[:templates])
+      end
+      nil
+    end
+  end
+
+  # Internal: Scan the specified directory for template files matching pattern and instantiate
+  # a Tilt template for each matched file.
+  #
+  # Returns the scan result as a [Hash]
+  def scan_dir template_dir, pattern, template_cache = nil
+    result, helpers = {}, nil
+    # Grab the files in the top level of the directory (do not recurse)
+    ::Dir.glob(pattern).select {|match| ::File.file? match }.each do |file|
+      if (basename = ::File.basename file) == 'helpers.js'
+        helpers = file
+        next
+      elsif (path_segments = basename.split '.').size < 2
+        next
+      end
+      if (name = path_segments[0]) == 'block_ruler'
+        name = 'thematic_break'
+      elsif name.start_with? 'block_'
+        name = name.slice 6, name.length
+      end
+      unless template_cache && (template = template_cache[file])
+        extsym = path_segments[-1].to_sym
+        case extsym
+        when :pug
+          unless @active_engines[extsym]
+            %x{
+              try {
+               const pug = require('pug')
+               template = { render: pug.compileFile(file), '$file': function() { return file } }
+              }
+              catch (e) {
+               throw #{IOError.new 'Unable to require the module \'pug\', please make sure that the module is installed.'}
+              }
+            }
+            @active_engines[extsym] = true
+          end
+        when :js
+          template = `{ render: require(file), '$file': function() { return file } }`
+        else
+          next
+        end
+      end
+      result[name] = template
+    end
+    if helpers || ::File.file?(helpers = %(#{template_dir}/helpers.js))
+      `require(helpers)`
```

Comment:
I think this feature is useless because the helper functions won't be available in the templates.
We could probably add an "helpers" property to the context?

```rb
@helpers = `require(helpers)`
# ...
`template.render({node: node, opts: opts, helpers: this.helpers})`
```